### PR TITLE
test:Sync functional test impl

### LINKED ISSUE #795: Sync: Write the unit (most), functional (some) and end-to-end (few) tests which verify the sync contracts
### Is your feature request related to a problem? Please describe.

See #793 

This task is to cover the work involved in verifying the individual contracts that, when strung together, make up the overall sync contract. (Definition of those contracts is covered by #794)

### Describe the solution you'd like

Verify the contracts by implementing the test cases for
- [x] (Client) How items are added to the 'uncommitted' queue on the client side (upon data store operations)
- [x] (Client) How the client processes that uncommitted queue (while sending updates to server) - e.g. how is the queue ordered, how is it de-duped, etc
- [x] (Server) How the server processes updates from the client (including the responses it generates) and what the expectations are - i.e. can we reject? what happens when we reject? and more
- [x] (Server) How items are added to the commit log on the server such that they are available for sync to the clients
- [x] (Server) How the server processes that log (when sending updates to client) - e.g. again ordering, de-duping, etc
- [x] (Client) How the client processes updates from the server - can the client reject? under what conditions? what happens upon a rejection?
- [x] The 7th contract is the contract for precisely how the client and server exchange information. Currently this is implemented using the batch verb and has state like 'syncInProgress' and 'isInSync'.
- [x] PR review and merge to trunk

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by purnimavenkatasubbu**
**- What I did**
Added functional tests for sync scenarios

**- How I did it**
Added tests for 

- sync with regex - by passing the regex in the preference
-  onDoneCallback
-  local key should not be synced to the cloud.

**- How to verify it**
at_client_sync_test.dart should pass

**- Description for the changelog**

### CODE REVIEW

**Comment by gkc** [comment]
@purnimavenkatasubbu I'd like to review this and get it merged. Can you merge trunk into this branch and resolve conflicts please?

**Comment by gkc** [comment]
@purnimavenkatasubbu There's a conflict again now. Please resolve - when it is ready for review again, please request me to review

File: tests/at_functional_test/test/at_client_sync_test.dart [hunk 1]
```diff
@@ -80,6 +80,7 @@ void main() {
     expect(updateResponse.isNotEmpty, true);
     // waiting for 15 seconds for sync to complete.
     await Future.delayed(Duration(seconds: 10));
+    atClientManager.atClient.syncService.sync();
```
Comment (gkc) [inline tests/at_functional_test/test/at_client_sync_test.dart:82]: Seems like this call to sync() should be done **before** doing the wait for 10 seconds

File: tests/at_functional_test/test/at_client_sync_test.dart [hunk 2]
```diff
@@ -89,10 +90,212 @@ void main() {
...
+    expect(
+        transformedValue.metadata!.updatedAt!.isBefore(DateTime.now()), true);
+  });
+}
+
+void onDoneCallback(syncResult) {
+  print('sync result $syncResult');
+  //always assert that the sync is successful when this method is triggered
+  expect(syncResult.syncStatus, SyncStatus.success);
+  //when this method is triggered always switch state to indicate that sync has been successful
```
Comment (gkc) [inline tests/at_functional_test/test/at_client_sync_test.dart]: there's no code which does what this comment says is happening ... ?

**Review by gkc — CHANGES_REQUESTED** [review]

File: tests/at_functional_test/test/at_client_sync_test.dart (hunk 2 above)
Comment (purnimavenkatasubbu) [reply tests/at_functional_test/test/at_client_sync_test.dart]: Updated

File: tests/at_functional_test/test/at_client_sync_test.dart (hunk 1 above)
Comment (purnimavenkatasubbu) [reply tests/at_functional_test/test/at_client_sync_test.dart:82]: Updated

**Review by gkc — APPROVED** [review]

**Pull request merged**